File: rules/java/tests/test_deserialization_tn.java
// True negative: Jackson JSON deserialization (safe)
import com.fasterxml.jackson.databind.ObjectMapper;
import javax.servlet.http.HttpServletRequest;

public class DataImporter {
    public MyDto importData(HttpServletRequest req) throws Exception {
        // ok: ez-spring-insecure-deserialization
        ObjectMapper mapper = new ObjectMapper();
        return mapper.readValue(req.getInputStream(), MyDto.class);
    }
}

File: rules/php/tests/test_deserialization_tn.php
<?php
// True negative: json_decode is safe
class DataController extends Controller
{
    public function import(Request $request)
    {
        // ok: ez-laravel-insecure-deserialization
        $obj = json_decode($request->input('data'), true);
        return response()->json($obj);
    }
}
